### ISSUE TITLE
Time for GEMMA 0.97 ?

### ISSUE THREAD

**Issue by tseemann** [genetics-statistics/GEMMA#90]
It's been 150 commits since.  Is it getting close to stable?

**Comment by pjotrp** [comment]
It is stable. Just need to check these really. https://github.com/genetics-statistics/GEMMA/milestones

**Comment by pjotrp** [comment]
I have made a preview release available on 
https://github.com/genetics-statistics/GEMMA/tree/gemma-0.97-preview

The binary can be installed and downloaded from http://test-gn2.genenetwork.org/ipfs/QmPuHFqYfBrRtdjpozMdzJj7RBN1AQEWkGeWwGDoLvWMeH with

```
   1. Download and unpack the tarball, e.g.:

        wget http://test-gn2.genenetwork.org/ipfs/QmPuHFqYfBrRtdjpozMdzJj7RBN1AQEWkGeWwGDoLvWMeH/y5i5ajpzkx8cb7fmsni7wa7jm82qh97j-gemma-git-gn2-0.97-pre3-a96b717-x86_64.tar.bz2
        mkdir tmp
        cd tmp
        tar xvjf ../y5i5ajpzkx8cb7fmsni7wa7jm82qh97j-gemma-git-gn2-0.97-pre3-a96b717-x86_64.tar.bz2

   2. Run the installer with a target directory

        ./install.sh ~/opt/gemma

   3. Run the software

        ~/opt/gemma/bin/gemma
```

@xiangzhou, @pcarbo please confirm that this works for you before we do a full release.

**Comment by pcarbo** [comment]
@pjotrp I ran this command on CentOS 7.x (Scientific Linux 7.x) and got the following error:

```bash
./install.sh /scratch/midway2/pcarbo/gemma
# gnu-install-bin 0.2.1 Copyright (C) 2017 Pjotr Prins <pjotr.prins@thebird.nl> and the GNU Guix project.
#     See also https://gitlab.com/pjotrp/gnu-install-bin
# Error: no implicit conversion of nil into String
# Done
```

This is current structure of the target directory:

```
gcc-5.4.0-lib-3x53yv4v144c9
gemma-git-gn2-0.97-pre3-a96b717-y5i5ajpzkx8cb
gfortran-5.4.0-lib-15plffwjdv8ii
glibc-2.25-n6nvxlk2j8ysf
gnu
gsl-2.4-as8vm640284nj
lapack-3.7.1-dm89yx4ws5akx
openblas-0.2.19-yw0hh910j9cfc
zlib-1.2.11-sfx1wh27i6gsr
```

Any ideas?

**Comment by pcarbo** [comment]
@pjotrp Also, when running on a MacBook Pro with Mac OS 10.12.6, I got the following error:

```bash
./install.sh ~/Desktop/gemma
# ./installer/bin/traveling-ruby-20141215-2.1.5-linux-x86_64/bin/ruby: line 6: /Users/pcarbo/Desktop/temp/installer/bin/traveling-ruby-20141215-2.1.5-linux-x86_64/bin.real/ruby: cannot execute binary file
# ./installer/bin/traveling-ruby-20141215-2.1.5-linux-x86_64/bin/ruby: line 6: /Users/pcarbo/Desktop/temp/installer/bin/traveling-ruby-20141215-2.1.5-linux-x86_64/bin.real/ruby: Undefined error: 0
# Done
```

**Comment by pcarbo** [comment]
@pjotrp How important is it to have `<regex>`? I see it is only used in `io.cpp`. It seems that it is not well supported by gcc 4.8.5, which is the default gcc installed on recent Linux (e.g., RedHat, CentOS) setups. This might introduce additional complications.

**Comment by pjotrp** [comment]
@pcarbo the installer is for Linux only.

Can you run the installer with -v and --debug switches, so I can tell where it is going wrong. You should be able to run gemma already by running

        /scratch/midway2/pcarbo/gemma/gemma-git-gn2-0.97-pre3-a96b717-y5i5ajpzkx8cb/bin/gemma

**Comment by pcarbo** [comment]
@pjotrp The installer worked this time. Did you fix any bugs? This is the console output I got:

```
gnu-install-bin 0.2.1 Copyright (C) 2017 Pjotr Prins <pjotr.prins@thebird.nl> and the GNU Guix project.
    See also https://gitlab.com/pjotrp/gnu-install-bin
  Installation to /scratch/midway2/pcarbo/gemma completed! Add the binaries to
  the path with:
    export PATH=/scratch/midway2/pcarbo/gemma/bin:$PATH
  The following binaries are available:
/scratch/midway2/pcarbo/gemma/bin/gemma
Done
```

**Comment by pcarbo** [comment]
*Note:* I was able to compile and run GEMMA successfully on my Mac (MacBook Pro, MacOS 10.12.6, gcc Apple LLVM version 8.1.0, GSL 2.4).

I'm having trouble compiling on Linux, I believe, due to the regex library issue [I mentioned above](https://github.com/genetics-statistics/GEMMA/issues/90#issuecomment-340075019).

**Comment by pjotrp** [comment]
Cool. I hope you can test it on some datasets before we release 0.97.

On regex - it is much safer than using strtok (for one strtok is not re-entrant). https://stackoverflow.com/questions/5999418/why-is-strtok-considered-unsafe. I created strtok_safe  after using the regex, so I can revert on that if it is *really* necessary. We should get far fewer segfaults from now on...

We still support gcc 4.9. I would prefer to retain it and we'll add more regex's rather than less. For those on older Linux distro's they can use that binary builds that we provide. GNU Guix can also be used to build on HPC. Otherwise 'find a workaround' would  be my suggestion. Build it in a VM and move it on the cluster. Conda may also be an option once we support it.

**Comment by pjotrp** [comment]
All issues are closed for 0.97. I am going to do a final comparison in the coming days while people test pre5. See https://groups.google.com/forum/#!topic/gemma-discussion/xNO72G-2Xl0. 

I'll try to build a binary version with an optimized OpenBlas for Linux.

**Comment by pjotrp** [comment]
https://groups.google.com/forum/#!topic/gemma-discussion/2NPWKhh3ixE
